Conditionally Target NET 9

FuPingFranco opened the pull request:
Add a net 9 target to Id Web with conditional compilation.

Successful Builds:

TargetNet9 = True (https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1281414&view=results)

TargetNet9 = False (https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1281375&view=results)

FuPingFranco commented on tests/E2E Tests/WebAppUiTests/WebAppUiTests.csproj:
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFrameworks>net8.0</TargetFrameworks>
+    <!--<TargetFrameworks Condition="'$(TargetNet9)'== 'True'">$(TargetFrameworks); net9.0</TargetFrameworks>-->
```
Seeing issues when targeting Net 9 potentially due to Daemon apps.

keegan-caruso commented on src/Microsoft.Identity.Web.GraphServiceClientBeta/Microsoft.Identity.Web.GraphServiceClientBeta.csproj:
```diff
@@ -17,7 +17,12 @@
-  <ItemGroup Condition="'$(TargetFramework)' != 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net9.0'">
+    <PackageReference Include="Microsoft.Graph.Beta" Version="5.56.0-preview" />
+    <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="9.0.0-preview.2.24128.5" />
```
9.0.0-preview.2.24128.5 is in so many places. When we want to update to the nice preview, it would be nice to have just a single var to update

pmaytak replied:
Should we use wildcard - 9.0.0-* - so it's easier to be up-to-date (don't have to update each time a new preview is released) and may catch some bugs/behavior changes in .NET?

jennyf19 replied:
Last year, that did not work, we had to specify the version, especially for the builds. :(

jennyf19 replied:
New previews are monthly, so it's doable, but not ideal

FuPingFranco replied:
Added var to control preview versions. Thanks Keegan, Peter and Jenny!

keegan-caruso commented on src/Microsoft.Identity.Web.TokenAcquisition/Microsoft.Identity.Web.TokenAcquisition.csproj:
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0' or '$(TargetFramework)' == 'net9.0'">
```
net6 or greater?

FuPingFranco replied:
Changed, Thanks Keegan!

westin-m commented on src/Microsoft.Identity.Web.TokenAcquisition/Microsoft.Identity.Web.TokenAcquisition.csproj:
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0' or '$(TargetFramework)' == 'net9.0'">
```
```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == 'NET6_0_OR_GREATER'">
```
Would this be better?

FuPingFranco replied:
Yeap, Thanks Westin!

westin-m commented on src/Microsoft.Identity.Web.TokenCache/Microsoft.Identity.Web.TokenCache.csproj:
```diff
@@ -11,7 +11,7 @@
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net462' Or '$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net462' Or '$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' Or '$(TargetFramework)' == 'net8.0' Or '$(TargetFramework)' == 'net9.0'">
```
```suggestion
  <ItemGroup Condition="'$(TargetFramework)' == 'net462' Or '$(TargetFrameworkIdentifier)' == 'NET6_0_OR_GREATER'">
```
Similar to the other one, would this be better?

FuPingFranco replied:
Yeap, Thanks Westin!

westin-m commented on tests/Microsoft.Identity.Web.Test/Microsoft.Identity.Web.Test.csproj:
```diff
@@ -9,7 +9,7 @@
   <ItemGroup>
     <ProjectReference Include="..\..\src\Microsoft.Identity.Web.DownstreamApi\Microsoft.Identity.Web.DownstreamApi.csproj" />
-    <ProjectReference Condition="'$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'" Include="..\..\src\Microsoft.Identity.Web.MicrosoftGraph\Microsoft.Identity.Web.MicrosoftGraph.csproj" />
+    <ProjectReference Condition="'$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' Or '$(TargetFramework)' == 'net8.0' Or '$(TargetFramework)' == 'net9.0'" Include="..\..\src\Microsoft.Identity.Web.MicrosoftGraph\Microsoft.Identity.Web.MicrosoftGraph.csproj" />
```
```suggestion
    <ProjectReference Condition="'$(TargetFrameworkIdentifier)' == 'NET6_0_OR_GREATER'" Include="..\..\src\Microsoft.Identity.Web.MicrosoftGraph\Microsoft.Identity.Web.MicrosoftGraph.csproj" />
```

FuPingFranco replied:
Yeap, Thanks Westin!

jmprieur commented on .github/workflows/dotnetcore.yml:
```diff
@@ -57,6 +62,9 @@ jobs:
       run: dotnet test --no-restore --no-build Microsoft.Identity.Web.sln -f net8.0 -p:FROM_GITHUB_ACTION=true --configuration Release --filter "(FullyQualifiedName!~Microsoft.Identity.Web.Test.Integration)&(FullyQualifiedName!~WebAppUiTests)&(FullyQualifiedName=IntegrationTests)"
 
+    - name: Test with .NET 9.0.x
+      run: dotnet test --no-restore --no-build Microsoft.Identity.Web.sln -f net9.0 -p:FROM_GITHUB_ACTION=true --configuration Release --filter "(FullyQualifiedName!~Microsoft.Identity.Web.Test.Integration)&(FullyQualifiedName!~WebAppUiTests)&(FullyQualifiedName=IntegrationTests)"
```
You also need to add `-p:TargetNet9=True`

FuPingFranco replied:
Added. Thanks JM!

jmprieur reviewed (changes requested):
There is no Net9 target for Microsoft.Identity.Web.UI

jennyf19 commented on Directory.Build.props:
```diff
@@ -57,7 +58,7 @@
   </ItemGroup>
 
-  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(TargetFramework)' != 'net8.0'">
+  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(TargetFrameworkIdentifier)' != 'NET8_0_OR_GREATER'">
```
this doesn't seem to work as we would think it would. I would target each specifically for now.

FuPingFranco replied:
Reverted. Thanks Jenny!

jennyf19 commented on src/Microsoft.Identity.Web.TokenAcquisition/Microsoft.Identity.Web.TokenAcquisition.csproj:
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == 'NET6_0_OR_GREATER'">
```
this doesn't seem to work, please revert back to targeting them each individually

FuPingFranco replied:
Reverted. Thanks Jenny!

jennyf19 commented on src/Microsoft.Identity.Web.TokenCache/Microsoft.Identity.Web.TokenCache.csproj:
```diff
@@ -11,7 +11,7 @@
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net462' Or '$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net462' Or '$(TargetFrameworkIdentifier)' == 'NET6_0_OR_GREATER'">
```
same as previous comment. careful w/taking suggestions from reviewers and not checking the code first

FuPingFranco replied:
Will do moving forward. Thanks Jenny!

jennyf19 commented on tests/DevApps/Directory.Build.props:
```diff
@@ -2,6 +2,7 @@
+    <TargetFrameworks Condition="'$(TargetNet9)'== 'True'">$(TargetFrameworks); net9.0</TargetFrameworks>
     <UseWip>true</UseWip>
     <IsPackable>false</IsPackable>
     <LangVersion>11</LangVersion>
```
not this PR, but I thought we had moved everything to LangVersion 12 - fyi @kellyyangsong

jennyf19 reviewed (changes requested):
after addressing the comments, please run this command:
`dotnet msbuild .\Microsoft.Identity.Web.sln -p:v=m` and see if things are looking good.

FuPingFranco commented:
> after addressing the comments, please run this command: `dotnet msbuild .\Microsoft.Identity.Web.sln -p:v=m` and see if things are looking good.

I can only run `msbuild Microsoft.Identity.Web.sln /p:v=m /p:TargetNet9=True` at the moment since using `dotnet` bring the OWIN apps error. I ran the msbuild command and everything looks fine. I got a warning that I will address on a separate PR but not an issue to complete this PR.

keegan-caruso commented on src/Microsoft.Identity.Web.TokenAcquisition/Microsoft.Identity.Web.TokenAcquisition.csproj:
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net7.0' Or '$(TargetFramework)' == 'net8.0' Or '$(TargetFramework)' == 'net9.0'" >
```
You can see examples of this in Identity Model

FuPingFranco replied:
Sure, let me try this and run a build.

FuPingFranco replied:
Qq, '.NETCoreApp' also, includes .net5.0 would that be a concern?

keegan-caruso replied:
Not quite how it works.

It evaluates per TargetFramework.

FuPingFranco replied:
@keegan-caruso it worked. Thanks!

jennyf19 reviewed (approved):
:shipit:

jmprieur commented on Directory.Build.props:
```diff
@@ -57,7 +58,7 @@
   </ItemGroup>
 
-  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(TargetFramework)' != 'net8.0'">
+  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and (('$(TargetFramework)' != 'net8.0') or ('$(TargetFramework)' != 'net9.0'))">
```
Shouldn't it be an and?

`"'$(TargetFramework)' != 'net8.0') or ('$(TargetFramework)' != 'net9.0'))"` is equivalent to:

`"! ('$(TargetFramework)' == 'net8.0') and ('$(TargetFramework)' == 'net9.0'))"` 

which is always true, so not sure that's you wanted to express.

Reminder:
`not A or not B` equivalent to `not (A and B)`

jennyf19 replied:
good catch @jmprieur

FuPingFranco replied:
It should be. Thanks for catching it @jmprieur!

jmprieur commented on src/Microsoft.Identity.Web.GraphServiceClient/Microsoft.Identity.Web.GraphServiceClient.csproj:
```diff
@@ -17,7 +17,12 @@
 
-  <ItemGroup Condition="'$(TargetFramework)' != 'net8.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net9.0'">
+    <PackageReference Include="Microsoft.Graph" Version="5.40.0" />
```
Why not always 5.40.0? is there an incompatiblity?

FuPingFranco replied:
Not for Microsoft.Graph from net8 to net9 but there's incompatibility in System.Diagnostics.DiagnosticSource as for why net7 and below are different that's something that was probably found in the NET8 work so, I just followed the pattern.

jmprieur replied:
@jennyf19 shouldn't we move to Graph 5.40.0 for all FW?

jennyf19 replied:
Can we open a work item and do it as a separate PR?

FuPingFranco replied:
Create issue to follow up https://github.com/AzureAD/microsoft-identity-web/issues/2748. Thanks @jmprieur & @jennyf19

jmprieur commented on tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs:
```diff
@@ -880,7 +880,9 @@ public void PreventChangesInOpenIdConnectOptionsToBeOverlooked()
             int numberOfProperties = typeof(OpenIdConnectOptions).GetProperties(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance).Length;
-#if NET8_0_OR_GREATER
+#if NET9_0_OR_GREATER
+            int expectedNumberOfProperties = 61;
```
Do we also need to update the MergedOption method?

jennyf19 replied:
there is a new member missing in the MergedOptions which needs to be added. could be another PR, please open an issue and add it to our backlog.

FuPingFranco replied:
Created new issue in backlog https://github.com/AzureAD/microsoft-identity-web/issues/2747. Thanks @jennyf19 & @jmprieur

jmprieur reviewed (commented):
I left a few questions.

pmaytak commented on Directory.Build.props:
```diff
@@ -57,7 +58,7 @@
   </ItemGroup>
 
-  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(TargetFramework)' != 'net8.0'">
+  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and (('$(TargetFramework)' != 'net8.0') or ('$(TargetFramework)' != 'net9.0'))">
```
```suggestion
  <ItemGroup Label="Build Tools" Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(TargetFramework)' != 'net8.0' and '$(TargetFramework)' != 'net9.0'">
```

pmaytak commented on src/Microsoft.Identity.Web.TokenAcquisition/Microsoft.Identity.Web.TokenAcquisition.csproj:
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' or '$(TargetFramework)' == 'net8.0'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'" >
```
Interesting. Didn't know about this.

FuPingFranco replied:
I think this is something that was first introduced on Net8 work. https://learn.microsoft.com/en-us/dotnet/standard/frameworks
